--- a/kernel-module-imx-gpu-viv-src/hal/os/linux/kernel/platform/freescale/gc_hal_kernel_platform_imx.c	2019-12-04 08:44:41.778935526 +0800
+++ b/kernel-module-imx-gpu-viv-src/hal/os/linux/kernel/platform/freescale/gc_hal_kernel_platform_imx.c	2019-12-04 08:50:26.171105246 +0800
@@ -52,6 +52,9 @@
 *
 *****************************************************************************/
 
+/* Disable PM on PHYTEC BSP */
+#undef CONFIG_PM
+#undef CONFIG_PM_OPP
 
 #include "gc_hal_kernel_linux.h"
 #include "gc_hal_kernel_platform.h"
@@ -86,14 +89,13 @@
 #    include <mach/busfreq.h>
 #  elif LINUX_VERSION_CODE < KERNEL_VERSION(3, 14, 29)
 #    include <linux/busfreq-imx6.h>
-#    include <linux/reset.h>
 #  else
 #    include <linux/busfreq-imx.h>
-#    include <linux/reset.h>
 #  endif
 #endif
 
 #include <linux/clk.h>
+#include <linux/reset.h>
 
 #if defined(IMX8_SCU_CONTROL)
 #  include <soc/imx8/sc/sci.h>
@@ -1169,7 +1171,9 @@
     int core = 0;
     struct gpu_clk *imx_clk = NULL;
     struct imx_priv *priv = &imxPriv;
+#if defined(CONFIG_PM)
     struct device *pmdev_last = NULL;/*legacy gpu device entry for imx6*/
+#endif
     struct clk *clk_core_last = NULL;/*vg has same core clk as 2d */
 
     for (core = 0; core < gcdMAX_GPU_COUNT; core++) {
