From cf5f7cffc0777bf0ec03511ec8321504d4399141 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Sebastian=20Dr=C3=B6ge?= <sebastian@centricular.com>
Date: Fri, 25 Apr 2014 17:32:59 +0200
Subject: [PATCH 03/21] giobasesink: Implement handling of the SEEKING query

---
 gst/gio/gstgiobasesink.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/gst/gio/gstgiobasesink.c b/gst/gio/gstgiobasesink.c
index 71e67f234..8a34b7067 100644
--- a/gst/gio/gstgiobasesink.c
+++ b/gst/gio/gstgiobasesink.c
@@ -343,6 +343,15 @@ gst_gio_base_sink_query (GstBaseSink * bsink, GstQuery * query)
         return TRUE;
       }
       return FALSE;
+    case GST_QUERY_SEEKING:
+      gst_query_parse_seeking (query, &format, NULL, NULL, NULL);
+      if (format == GST_FORMAT_BYTES || format == GST_FORMAT_DEFAULT) {
+        gst_query_set_seeking (query, GST_FORMAT_BYTES,
+            GST_GIO_STREAM_IS_SEEKABLE (sink->stream), 0, -1);
+      } else {
+        gst_query_set_seeking (query, format, FALSE, 0, -1);
+      }
+      return TRUE;
     default:
       return GST_BASE_SINK_CLASS (parent_class)->query (bsink, query);
   }
-- 
2.19.0

