--- a/libc/sysdeps/linux/common/sigprocmask.c	2018-09-26 10:39:59.057945127 +0800
+++ a/libc/sysdeps/linux/common/sigprocmask.c	2018-09-26 10:41:14.014373971 +0800
@@ -23,7 +23,7 @@
 _syscall4(int, __rt_sigprocmask, int, how, const sigset_t *, set,
 		  sigset_t *, oldset, size_t, size)
 
-int sigprocmask(int how, const sigset_t * set, sigset_t * oldset)
+int __attribute__((weak)) sigprocmask(int how, const sigset_t * set, sigset_t * oldset)
 {
 #ifdef SIGCANCEL
 	sigset_t local_newmask;
@@ -58,7 +58,7 @@
 _syscall3(int, __syscall_sigprocmask, int, how, const sigset_t *, set,
 		  sigset_t *, oldset)
 
-int sigprocmask(int how, const sigset_t * set, sigset_t * oldset)
+int __attribute__((weak)) sigprocmask(int how, const sigset_t * set, sigset_t * oldset)
 {
 #ifdef SIGCANCEL
 	sigset_t local_newmask;
